File: java/test/Ice/adapterDeactivation/ServantLocatorI.java
// **********************************************************************
//
// Copyright (c) 2003-2014 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package test.Ice.adapterDeactivation;

import test.Ice.adapterDeactivation.Test.Cookie;

public final class ServantLocatorI implements Ice.ServantLocator
{
    public
    ServantLocatorI()
    {
        _destroyed = false;
    }

    protected synchronized void
    finalize()
        throws Throwable
    {
        test(_destroyed);
    }

    private static void
    test(boolean b)
    {
        if(!b)
        {
            throw new RuntimeException();
        }
    }

    public Ice.Object
    locate(Ice.Current current, Ice.LocalObjectHolder cookie)
    {
        synchronized(this)
        {
            test(!_destroyed);
        }

        test(current.id.category.length() == 0);
        test(current.id.name.equals("test"));

        cookie.value = new CookieI();

        return new TestI();
    }

    public void
    finished(Ice.Current current, Ice.Object servant, java.lang.Object cookie)
    {
        synchronized(this)
        {
            test(!_destroyed);
        }

        Cookie co = (Cookie)cookie;
        test(co.message().equals("blahblah"));
    }

    public synchronized void
    destroy(String category)
    {
        synchronized(this)
        {
            test(!_destroyed);

            _destroyed = true;
        }
    }

    private boolean _destroyed;
}

File: cs/test/Ice/adapterDeactivation/ServantLocatorI.cs
// **********************************************************************
//
// Copyright (c) 2003
// ZeroC, Inc.
// Billerica, MA, USA
//
// All Rights Reserved.
//
// Ice is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License version 2 as published by
// the Free Software Foundation.
//
// **********************************************************************

using System;

public sealed class ServantLocatorI : Ice.LocalObjectImpl, Ice.ServantLocator
{
    public ServantLocatorI()
    {
        _deactivated = false;
    }
    
    ~ServantLocatorI()
    {
        test(_deactivated);
    }
    
    private static void test(bool b)
    {
        if(!b)
        {
            throw new System.Exception();
        }
    }
    
    public Ice.Object locate(Ice.Current current, out Ice.LocalObject cookie)
    {
        test(!_deactivated);
        
        test(current.id.category.Length == 0);
        test(current.id.name.Equals("test"));
        
        cookie = new CookieI();
        
        return new TestI();
    }
    
    public void finished(Ice.Current current, Ice.Object servant, Ice.LocalObject cookie)
    {
        test(!_deactivated);
        
        Cookie co = (Cookie) cookie;
        test(co.message().Equals("blahblah"));
    }
    
    public void deactivate(string category)
    {
        test(!_deactivated);
        
        _deactivated = true;
    }
    
    private bool _deactivated;
}
